src/java.base/share/classes/sun/security/pkcs12/MacData.java

```diff
@@ -70,103 +94,218 @@ class MacData {
-        this.digestAlgorithmName = algid.getName();
-        this.digestAlgorithmParams = algid.getParameters();
-
-        if (digest == null) {
-            throw new NullPointerException("the digest " +
-                                           "parameter must be non-null");
-        } else if (digest.length == 0) {
-            throw new IllegalArgumentException("the digest " +
-                                                "parameter must not be empty");
+    /*
+     * Calculate MAC using HMAC algorithm (required for password integrity)
+     *
+     * Hash-based MAC algorithm combines secret key with message digest to
+     * create a message authentication code (MAC)
+     */
+    static byte[] calculateMac(char[] passwd, byte[] data,
+            String macAlgorithm, int macIterationCount, byte[] salt)
+            throws IOException, NoSuchAlgorithmException {
+        final PBEParameterSpec params;
+        String algName;
+        String kdfHmac;
+        String hmac;
+
+        macAlgorithm = macAlgorithm.toLowerCase(ENGLISH);
+        if (macAlgorithm.startsWith("pbewith")) {
+            algName = "PBMAC1";
+            kdfHmac = MacData.parseKdfHmac(macAlgorithm);
+            hmac = MacData.parseHmac(macAlgorithm);
+            if (hmac == null) {
+                hmac = kdfHmac;
+            }
+        } else if (macAlgorithm.startsWith("hmacpbe")) {
+            algName = macAlgorithm.substring(7);
+            kdfHmac = null;
+            hmac = macAlgorithm;
         } else {
-            this.digest = digest.clone();
+            throw new ParsingException("unexpected algorithm '"
+                    + macAlgorithm + "'");
         }
 
-        this.macSalt = salt;
-        this.iterations = iterations;
+        params = new PBEParameterSpec(salt, macIterationCount);
 
-        // delay the generation of ASN.1 encoding until
-        // getEncoded() is called
-        this.encoded = null;
+        try {
+            byte[] macResult = getMac(macAlgorithm, passwd, params, data,
+                    kdfHmac, hmac, -1);
 
+            DerOutputStream bytes = new DerOutputStream();
+            bytes.write(encode(algName, macResult, params, kdfHmac, hmac,
+                    macResult.length));
+            return bytes.toByteArray();
+        } catch (InvalidKeySpecException | InvalidKeyException |
+                    InvalidAlgorithmParameterException e) {
+            throw new IOException("calculateMac failed: " + e, e);
+        }
     }
 
-    String getDigestAlgName() {
-        return digestAlgorithmName;
+    String getMacAlgorithm() {
+        return this.macAlgorithm;
     }
 
     byte[] getSalt() {
-        return macSalt;
+        return this.macSalt;
     }
 
     int getIterations() {
-        return iterations;
-    }
-
-    byte[] getDigest() {
-        return digest;
+        return this.iterations;
     }
 
     /**
      * Returns the ASN.1 encoding of this object.
```

Comment:
This is a `static` method. There is no "this object".

Reply:
Duh! Fixed.